lib/Sema/DerivedConformanceParameterized.cpp

```diff
@@ -0,0 +1,393 @@
+  StructDecl *parametersDecl = nullptr;
+
+  for (auto memberDecl : nominal->getMembers()) {
+    auto structDecl = dyn_cast<StructDecl>(memberDecl);
+    if (!structDecl || structDecl->getName() != ctx.Id_Parameters)
+      continue;
+    parametersDecl = structDecl;
+    if (isValidParametersStruct(nominal, structDecl))
+      return std::make_pair(parametersDecl, Valid);
+    else
+      status = Invalid;
+  }
+  return std::make_pair(parametersDecl, status);
+}
+
+// Return the protocol requirement with the specified name.
+static ValueDecl *getProtocolRequirement(ProtocolDecl *proto, Identifier name) {
+  auto lookup = proto->lookupDirect(name);
+  assert(lookup.size() == 1 && "Ambiguous protocol requirement");
+  return lookup[0];
+}
+
+static void derivedBody_allParametersGetter(AbstractFunctionDecl *getterDecl) {
+  auto *nominal = getterDecl->getDeclContext()
+      ->getAsNominalTypeOrNominalTypeExtensionContext();
+  auto &C = nominal->getASTContext();
+
+  auto *parametersDecl = getParametersStructDecl(nominal).first;
+  assert(parametersDecl && "'Parameters' struct not found but should exist");
+  ConstructorDecl *parametersInitDecl = nullptr;
+  for (auto member : parametersDecl->getMembers()) {
+    auto initDecl = dyn_cast<ConstructorDecl>(member);
+    if (!initDecl || !initDecl->isMemberwiseInitializer())
+      continue;
+    assert(!parametersInitDecl && "'Parameters' initializer already found");
+    parametersInitDecl = initDecl;
+  }
+  assert(parametersInitDecl && "'Parameters' implicit initializer not found");
+
+  auto *selfDecl = getterDecl->getImplicitSelfDecl();
+  auto *selfDRE =
+      new (C) DeclRefExpr(selfDecl, DeclNameLoc(), /*Implicit*/ true);
+
+  auto *initDRE =
+      new (C) DeclRefExpr(parametersInitDecl, DeclNameLoc(), /*Implicit*/ true);
+  initDRE->setFunctionRefKind(FunctionRefKind::SingleApply);
+
+  auto parametersType =
+      nominal->mapTypeIntoContext(parametersDecl->getDeclaredInterfaceType());
+  Expr *baseExpr = TypeExpr::createImplicit(parametersType, C);
+  auto *initExpr = new (C) ConstructorRefCallExpr(initDRE, baseExpr);
+  initExpr->setThrows(false);
+  initExpr->setImplicit();
+
+  auto *parameterizedProto = C.getProtocol(KnownProtocolKind::Parameterized);
+  auto allParametersReq =
+    getProtocolRequirement(parameterizedProto, C.Id_allParameters);
+
+  auto getUnderlyingParameter = [&](Expr *expr, VarDecl *param) -> Expr * {
+    auto module = nominal->getModuleContext();
+    auto confRef =
+        module->lookupConformance(param->getType(), parameterizedProto);
+    if (!confRef)
+      return expr;
+    auto conf = confRef->getConcrete();
+    auto allParamsDecl = conf->getWitnessDecl(allParametersReq, nullptr);
+    return new (C) MemberRefExpr(expr, SourceLoc(), allParamsDecl,
+                                 DeclNameLoc(), /*Implicit*/ true);
+  };
+
+  SmallVector<Expr *, 2> members;
+  SmallVector<Identifier, 2> memberNames;
```

Comment:
Use one vector of tuples instead of two vectors. That statically guarantees equal number of members and member names.

Reply:
All `CallExpr::create` methods take `ArrayRef<Expr *> args` and `ArrayRef<Identifier> argLabels` separately, however.

Reply:
Oh ok, I didn't notice the use sites.